Automattic/woocommerce-payments
# Fix refunding of orders without _payment_method_id

Fixes #4297 

#### Changes proposed in this Pull Request

This PR fixes the required param type error when refunding orders that do not have a `_payment_method_id` set in their meta.

The missing meta could be caused when a payment intent succeeded webhook arrives before the UPE payment redirect is processed. When that happens, the order cannot be refunded from the WooCommerce dashboard.

It also fixes a receipt_number typo in refund tests and moves all refund tests to a separate test file to reduce the clutter in `WC_Payment_Gateway_WCPay_Test`.

*Note: Sorry for the noisy tests move. The changes from https://github.com/Automattic/woocommerce-payments/pull/4314/commits/201506bb1d2237cb6f762e62e50c2d70fc599b00 are just moving the tests to another file, without any changes to them. Please check https://github.com/Automattic/woocommerce-payments/pull/4314/commits/58386e7909175841b5660eeb1d88bd27e50450fe to see the main changes that fix this issue.*

#### Testing instructions

*Before running these test instructions with this PR, attempt to reproduce the bug on `develop`.*

* Process a WC Pay order and delete its `_payment_method_id` from the `wp_postmeta` DB table
* Navigate to the order page and attempt to refund it
    * If using `develop`, the refund should fail with the "Error: Missing required param: type." message
    * If using this branch, the refund should be successful
* While using this PR's branch
    * Make sure that refunding an order that has a right `_payment_method_id` meta works
    * Make sure that refunding an order that has an empty string (`''`) `_payment_method_id` meta works
    * Make sure that refunding an order that has a `NULL` `_payment_method_id` meta works

-------------------


**Post merge**
